### ISSUE TITLE
(t) resolve outstanding test discrepancy

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2408]
Current testing channel (post merge of #2407) has an observed discrepancy across 2 similar install instances:
## Leap 15.3-x86_64 (KVM instance on mobile Haswell host)
```
Ran 229 tests in 29.565s

OK
```

## Leap 15.3-aarch64 (KVM instance on Pi4 host)
```
======================================================================
FAIL: test_post_status_running_cli_balance (storageadmin.tests.test_pool_balance.PoolBalanceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_pool_balance.py", line 242, in test_post_status_running_cli_balance
    self.assertEqual(r.data["status"], bstatus["status"], msg=e_msg)
AssertionError: failed recording cli status. 
 Got ({'status': u'unknown', 'start_time': '2022-10-28T12:45:23.699534+01:00', 'internal': False, 'end_time': None, 'tid': None, 'message': u'Suspected cli balance detected. New entry created.', 'percent_done': 50, u'id': 3, 'pool': 2}) 
Expected ({'status': 'finished', 'percent_done': 100}).
Pool name test-pool-balance, Test data index 4

======================================================================
FAIL: test_delete_requests (storageadmin.tests.test_snapshot.SnapshotTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_snapshot.py", line 274, in test_delete_requests
    status.HTTP_200_OK, msg=response.data)
AssertionError: ["Error running a command. cmd = /usr/bin/umount -l /mnt2/share2/.snap2. rc = 32. stdout = ['']. stderr = ['umount: /mnt2/share2/.snap2: not mounted.', '']", 'Traceback (most recent call last):\n  File "/opt/rockstor-dev/src/rockstor/rest_framework_custom/generic_view.py", line 41, in _handle_exception\n    yield\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/snapshot.py", line 291, in delete\n    self._delete_snapshot(request, sid, snap_name=snap_name)\n  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/utils/decorators.py", line 185, in inner\n    return func(*args, **kwargs)\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/snapshot.py", line 272, in _delete_snapshot\n    share, snapshot.real_name, snapshot.qgroup, on=False\n  File "/opt/rockstor-dev/eggs/Django-1.11.29-py2.7.egg/django/utils/decorators.py", line 185, in inner\n    return func(*args, **kwargs)\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/snapshot.py", line 117, in _toggle_visibility\n    umount_root(snap_mnt_pt)\n  File "/opt/rockstor-dev/src/rockstor/fs/btrfs.py", line 602, in umount_root\n    raise ce\nCommandException: Error running a command. cmd = /usr/bin/umount -l /mnt2/share2/.snap2. rc = 32. stdout = [\'\']. stderr = [\'umount: /mnt2/share2/.snap2: not mounted.\', \'\']\n']

----------------------------------------------------------------------
Ran 229 tests in 66.466s

FAILED (failures=2)
```
Where the later failures are in both an old and a new test. However given our recent revamp of most of our existing tests, in pr #2400, this may well relate to an unintended sensitivity to machine/fs state/speed as the aarch64 does have, in this case, some snapshots that the x86_64 does not, and also works far slower.

**Comment by FroggyFlox** [comment]
Hi @phillxnet,

I just tried running all tests on a fresh source build of the current testing branch on my x86_64 KVM and I do have one failure, the same as the first failure you report on Pi4:
```
======================================================================
FAIL: test_post_status_running_cli_balance (storageadmin.tests.test_pool_balance.PoolBalanceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_pool_balance.py", line 242, in test_post_status_running_cli_balance
    self.assertEqual(r.data["status"], bstatus["status"], msg=e_msg)
AssertionError: failed recording cli status. 
 Got ({'status': u'unknown', 'start_time': '2022-10-28T11:48:33.504549-04:00', 'internal': False, 'end_time': None, 'tid': None, 'message': u'Suspected cli balance detected. New entry created.', 'percent_done': 50, u'id': 3, 'pool': 2}) 
Expected ({'status': 'finished', 'percent_done': 100}).
Pool name test-pool-balance, Test data index 4

----------------------------------------------------------------------
Ran 229 tests in 28.184s
```

**Comment by FroggyFlox** [comment]
I just noticed that the `start_time` of the returned balance status is the current machine time so it seems indeed we are triggering a real balance here and not mocking one?

**Comment by phillxnet** [comment]
@FroggyFlox Thanks for taking a look; I'll look to that one next. From memory I had thought we were mocking the balance command itself to be a shell null opp! However the start time does default, in db, to be when the task was started (again from memory). I've just been looking at the 'other' failure here and we are simply missing a mock that is required if the mount point exists, so we have a fix for that in the pending pr I'm working on for this issue:

>         cls.patch_umount_root = patch("storageadmin.views.snapshot.umount_root")
>         cls.mock_umount_root = cls.patch_umount_root.start()
>         cls.mock_umount_root.return_value = "out", "err", 0

results in an apparent fix with the reproducer instance.

> ./bin/test -v 2 -p test_snapshot.py
> 
> test_clone_command (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
> test_delete_requests (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
> test_get (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
> test_post_requests_1 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
> test_post_requests_2 (storageadmin.tests.test_snapshot.SnapshotTests) ... ok
> 
> ----------------------------------------------------------------------
> Ran 5 tests in 0.604s

Got called away today so unfortunately I didn't make the progress I'd hoped with this. But all looking completely doable: just wanted to be sure we didn't have an obscure arch issue or something.

**Comment by FroggyFlox** [comment]
No problem; I just thought I would share in case it helps.


> From memory I had thought we were mocking the balance command itself to be a shell null opp!

I think we do mock with a return value of True.


> However the start time does default, in db, to be when the task was started (again from memory)

Oh, so we could have a current start_time even with a properly mocked balance start... I didn't think of that, sorry.

**Comment by phillxnet** [comment]
@FroggyFlox Thanks for the pointer on your x86_64 instance also failing, My clean source install on the same passes. But this looks again like an insufficient mocking thing where there may be inadvertent sensitivity to system state. Popping in some log notes here for comparison: running only the failing test (remarked out the others for now):
## passing instance
```
./bin/test -v 2 -p test_pool_balance.py
...
test_post_status_running_cli_balance (storageadmin.tests.test_pool_balance.PoolBalanceTests) ... running test data index 0
running test data index 1
running test data index 2
running test data index 3
running test data index 4
running test data index 0
running test data index 1
running test data index 2
running test data index 3
running test data index 4
ok

----------------------------------------------------------------------
Ran 1 test in 0.447s

OK
```
### Associated debug log:
```
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (False), Internal: (False), Live Status: ({'status': u'finished', 'percent_done': 100})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = 1d9aa3bc-last-poolbalance-in-fixture Pool (test-pool-balance) End_time (2022-07-05 10:57:23.170000+00:00)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:194] Matching task result = None
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:212] PROPOSED STATUS = MODEL STATUS: NO UPDATE REQUIRED
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 6})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = 1d9aa3bc-last-poolbalance-in-fixture Pool (test-pool-balance) End_time (2022-07-05 10:57:23.170000+00:00)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:194] Matching task result = None
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': 'running', 'percent_done': 6}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': 'running', 'percent_done': 6}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:224] ++++++ Suspected cli balance detected. New entry created. Pool name test-pool-balance, PoolBalance ID 3.
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 6})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (test-pool-balance) End_time (None)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:212] PROPOSED STATUS = MODEL STATUS: NO UPDATE REQUIRED
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 50})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (test-pool-balance) End_time (None)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'running', 'percent_done': 6}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': 'running', 'percent_done': 50}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': 'running', 'percent_done': 50}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:228] UPDATING RECORD ID 3 WITH {'status': 'running', 'percent_done': 50}.
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (False), Internal: (False), Live Status: ({'status': u'finished', 'percent_done': 100})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (test-pool-balance) End_time (None)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'running', 'percent_done': 50}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:228] UPDATING RECORD ID 3 WITH {'status': u'finished', 'percent_done': 100}.
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (False), Internal: (False), Live Status: ({'status': u'finished', 'percent_done': 100})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:251] ====== Returning Empty as no PoolBalance - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 6})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:165] ++++++ Suspected cli balance detected. New entry created. Pool name ROOT, PoolBalance ID 4.
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 6})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (ROOT) End_time (None)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:212] PROPOSED STATUS = MODEL STATUS: NO UPDATE REQUIRED
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 50})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (ROOT) End_time (None)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'running', 'percent_done': 6}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': 'running', 'percent_done': 50}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': 'running', 'percent_done': 50}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:228] UPDATING RECORD ID 4 WITH {'status': 'running', 'percent_done': 50}.
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:46:00] DEBUG [fs.btrfs:2004] Balance active: (False), Internal: (False), Live Status: ({'status': u'finished', 'percent_done': 100})
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (ROOT) End_time (None)
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'running', 'percent_done': 50}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:228] UPDATING RECORD ID 4 WITH {'status': u'finished', 'percent_done': 100}.
[29/Oct/2022 12:46:00] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
```
## failing instance:
```
rleap15-3aarch64:/opt/rockstor-dev # ./bin/test -v 2 -p test_pool_balance.py
...
test_post_status_running_cli_balance (storageadmin.tests.test_pool_balance.PoolBalanceTests) ... running test data index 0
running test data index 1
running test data index 2
running test data index 3
running test data index 4
FAIL

======================================================================
FAIL: test_post_status_running_cli_balance (storageadmin.tests.test_pool_balance.PoolBalanceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_pool_balance.py", line 244, in test_post_status_running_cli_balance
    self.assertEqual(r.data["status"], bstatus["status"], msg=e_msg)
AssertionError: failed recording cli status. 
 Got ({'status': u'unknown', 'start_time': '2022-10-29T12:38:51.974197+01:00', 'internal': False, 'end_time': None, 'tid': None, 'message': u'Suspected cli balance detected. New entry created.', 'percent_done': 50, u'id': 3, 'pool': 2}) 
Expected ({'status': 'finished', 'percent_done': 100}).
Pool name test-pool-balance, Test data index 4

----------------------------------------------------------------------
Ran 1 test in 0.752s

FAILED (failures=1)

```
### Failing instance debug log:
```
[29/Oct/2022 12:38:51] INFO [system.pkg_mgmt:118] No "rockstor" package found: source install?
[29/Oct/2022 12:38:51] DEBUG [fs.btrfs:2004] Balance active: (False), Internal: (False), Live Status: ({'status': u'unknown'})
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = 1d9aa3bc-last-poolbalance-in-fixture Pool (test-pool-balance) End_time (2022-07-05 10:57:23.170000+00:00)
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:194] Matching task result = None
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': u'unknown'}
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': u'unknown'}
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:38:51] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 6})
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = 1d9aa3bc-last-poolbalance-in-fixture Pool (test-pool-balance) End_time (2022-07-05 10:57:23.170000+00:00)
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:194] Matching task result = None
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'finished', 'percent_done': 100}
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': 'running', 'percent_done': 6}
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': 'running', 'percent_done': 6}
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:224] ++++++ Suspected cli balance detected. New entry created. Pool name test-pool-balance, PoolBalance ID 3.
[29/Oct/2022 12:38:51] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:38:52] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 6})
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (test-pool-balance) End_time (None)
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:212] PROPOSED STATUS = MODEL STATUS: NO UPDATE REQUIRED
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:38:52] DEBUG [fs.btrfs:2004] Balance active: (True), Internal: (False), Live Status: ({'status': 'running', 'percent_done': 50})
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (test-pool-balance) End_time (None)
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'running', 'percent_done': 6}
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': 'running', 'percent_done': 50}
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': 'running', 'percent_done': 50}
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:228] UPDATING RECORD ID 3 WITH {'status': 'running', 'percent_done': 50}.
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
[29/Oct/2022 12:38:52] DEBUG [fs.btrfs:2004] Balance active: (False), Internal: (False), Live Status: ({'status': u'unknown'})
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:171] Latest PoolBalance: TID = None Pool (test-pool-balance) End_time (None)
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:50] PRIOR_STATUS = {'status': u'running', 'percent_done': 50}
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:51] LIVE_STATUS = {'status': u'unknown'}
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:66] PROPOSED STATUS = {'status': u'unknown'}
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:228] UPDATING RECORD ID 3 WITH {'status': u'unknown'}.
[29/Oct/2022 12:38:52] DEBUG [storageadmin.views.pool_balance:246] ====== Returning PoolBalance serialized - status command. ======
```
So we have a difference in mocking here !! working is as intended reported as 
finnished
running
running
running
finished
But in the failing instance we have a bunch of unknowns.

**Comment by phillxnet** [comment]
We have an un-mocked response from our system regarding prior 'internal' balance:
Additional debug logging added, first and last call from test-pool-balance target pool loop within test at question.

## Passing system:
```
[29/Oct/2022 14:56:49] DEBUG [fs.btrfs:1997] balance_status_all(test-pool-balance) getting balance_status() of ({'status': 'finished', 'percent_done': 100})
[29/Oct/2022 14:56:49] DEBUG [fs.btrfs:2005] balance_status_all(test-pool-balance) getting balance_status_internal() of ({'status': u'finished', 'percent_done': 100})
[29/Oct/2022 14:56:49] DEBUG [fs.btrfs:2014] Balance active: (False), Internal: (False), Live Status: ({'status': u'finished', 'percent_done': 100})
...
[29/Oct/2022 14:56:49] DEBUG [fs.btrfs:1997] balance_status_all(test-pool-balance) getting balance_status() of ({'status': 'finished', 'percent_done': 100})
[29/Oct/2022 14:56:49] DEBUG [fs.btrfs:2005] balance_status_all(test-pool-balance) getting balance_status_internal() of ({'status': u'finished', 'percent_done': 100})
[29/Oct/2022 14:56:49] DEBUG [fs.btrfs:2014] Balance active: (False), Internal: (False), Live Status: ({'status': u'finished', 'percent_done': 100})
```

## Failing system:

```
[29/Oct/2022 15:01:57] DEBUG [fs.btrfs:1997] balance_status_all(test-pool-balance) getting balance_status() of ({'status': 'finished', 'percent_done': 100})
[29/Oct/2022 15:01:57] DEBUG [fs.btrfs:2005] balance_status_all(test-pool-balance) getting balance_status_internal() of ({'status': u'unknown'})
[29/Oct/2022 15:01:57] DEBUG [fs.btrfs:2014] Balance active: (False), Internal: (False), Live Status: ({'status': u'unknown'})
...
[29/Oct/2022 15:01:57] DEBUG [fs.btrfs:1997] balance_status_all(test-pool-balance) getting balance_status() of ({'status': 'finished', 'percent_done': 100})
[29/Oct/2022 15:01:57] DEBUG [fs.btrfs:2005] balance_status_all(test-pool-balance) getting balance_status_internal() of ({'status': u'unknown'})
[29/Oct/2022 15:01:57] DEBUG [fs.btrfs:2014] Balance active: (False), Internal: (False), Live Status: ({'status': u'unknown'})
```

So we have a failure to mock a recently more incorporated balance_status_internal() (into bool_balance's awareness) and pool_balance_all() added more recently, inadvertently prioritising an unknown internal state.

No working now on creating a test for our newly added and likely suspect balance_status_all() to reproduce and then help correct what looks to be not the behaviour we expect from our overview balance status mechanism.

**Comment by phillxnet** [comment]
Our mocking of mount_root() without mocking balance_status_internal() causes the latter to return an unknown status, but only if the now mocked mount point concerned doesn't exist. If it does, as in my 'passing' instance we get a more sane return of what we expect but from the host pool mount point (/mnt2/test-pool-balance).

So we have a missing mock of balance_status_internal() (or alternatively we mock balance_status_all())  in our test_pool_balance.py which in turn shows we are in the newly added balance_status_all(), in some settings, prioritising an unknown state (from balance_status_internal()) over a known one from balance_status():

- getting balance_status() of ({'status': 'finished', 'percent_done': 100})
- getting balance_status_internal() of ({'status': u'unknown'})
-- resulting in the (partially wrong) Balance active: (False), Internal: (False), Live Status: ({'status': u'unknown'})

The more sane response in this case should be:
-- resulting in the Balance active: (False), Internal: (False), Live Status: ({'status': u'finished'})
As unknown is just that, we can only infer if there is an internal balance. If we don't know we must assume there isn't one, or go with what balance_status() tell us. In time we hope to see "btrfs balance status mnt_pt" be aware of internal balances. Until that is the case we need to carry our own inference on this volume state.

All of which, in turn, throws our higher-ups. If balance_status_internal doesn't know, (unknown) we should not use/prioritise it's status data. It can only tell us if there is one, otherwise we go with the returned status of balance_status().

Additional tests are being developed to assert this behaviour while the reproducer is at-hand.

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2411
